### ISSUE TITLE
Support both Promise / Observable return types

### ISSUE THREAD

**Issue by vladimiry** [vladimiry/electron-rpc-api#1]
It will require preserving the Promise / Observable return type in runtime so the API structure will need to be defined as a lightweight value rather than a type/interface.

Currently looking API definition:
```typescript
export interface IpcMainApi {
    ping: ApiMethod<{ domain: string, times: number }, { domain: string, value: number }>;
    quitApp: ApiMethodNoArgument<never>;
}
```

Is going to look similar to this form:
```typescript
const apiDefinition = {
    ping: ActionType.Observable<[{ domain: string, times: number }], { domain: string, value: number }>();
    quitApp: ActionType.Promise();
}
```

The default generics arguments/output parameters of the `ActionType.Observable` / `ActionType.Promise` functions is going to look like `IN extends any[] = []` and `OUT extends any = void`. So `ActionType.Promise<[], void>()` is the same as `ActionType.Promise()`.

The change is going to be released with v5. Work is in progress.

**Comment by vladimiry** [comment]
I think it's better not to wrap the function argument to the array as we get some benefits not doing that:
- API definition is less verbose as there is no need to bracket arguments/input generic parameter with `[]`. We can use `never / void` types to indicate that the action function doesn't have an argument.
- Wrapping all the action function input data to a single argument allows us to send second and more service purpose arguments. This is the module v1-v4/current versisons behavior. Like for example we can send second arg with `type Context<RawArgs extends any> = { rawArguments: RawArgs }` as a second argument where the `rawArguments` values is raw arguments passed to the listener function. So for @electron `rawArguments` would be the `Electron.IpcMessageEvent` type. The module actually already passes this `rawArguments` data to the function context but it's a more verbose way than passing as an argument since currently, it requires some manipulations. We can make `RawArgs` generic argument for registering API definition and so all the actions will get second argument type inferred automatically.

**Comment by vladimiry** [comment]
Resolved with [v5.0.0-beta1](https://github.com/vladimiry/electron-rpc-api/releases/tag/v5.0.0-beta1) release. So both Observable and Promise API endpoints are now supported, see usage example in the readme.
